Add new transitions report

tsteur opened the pull request:
As you can maybe see in the first commit I tried to first make it work as a report  with a visualization and including related report etc but couldn't get it to work. There were various challenges. So made it a widget instead. Was still quite tricky to get it to work and is a bit messy since it was all hard adjusted to be a popover.

Looks like this:
![image](https://user-images.githubusercontent.com/273120/45988614-e6ed3700-c0cb-11e8-91a7-b15e0a0c973a.png)

No data screen:
![image](https://user-images.githubusercontent.com/273120/45988628-f4a2bc80-c0cb-11e8-8266-af61f8add42e.png)

Menu: 
![image](https://user-images.githubusercontent.com/273120/45988640-008e7e80-c0cc-11e8-9c43-5793ef9dc666.png)

There is no search in the label field for now. It's only our regular select. Should do for now I'd say and at some point need to add support for autocomplete or so.

I'll try to add some UI tests if that's looking good @mattab 

fix #12865

tsteur committed:
trying to add a report through visualization

tsteur committed:
refs #12865 added transitions report

tsteur committed:
small tweak

diosmosis commented on plugins/Transitions/API.php:
```diff
@@ -617,4 +617,22 @@ protected function processTransitionsToFollowingPages($type, $actions)
+        $table->addRowFromArray(array(Row::COLUMNS => array('nb_visits' => 5)));
+
+        return $table;
+    }
```
Should this be here?

tsteur replied:
will remove it, was auto generated 👍

tsteur committed:
remove auto generated api

diosmosis commented:
Testing locally noticed these issues:

1) the report doesn't seem to be centered, might look better if it was

![image](https://user-images.githubusercontent.com/125140/49333573-1316af80-f576-11e8-931c-41ed44b65249.png)

2) If I click in the transitions report, I see a popup like this

![image](https://user-images.githubusercontent.com/125140/49333583-275aac80-f576-11e8-88af-c68cea470007.png)

  and then the report is empty

![image](https://user-images.githubusercontent.com/125140/49333585-317cab00-f576-11e8-9ce7-f2c960dfa0cb.png)

3) the select dropdown is under the report

![image](https://user-images.githubusercontent.com/125140/49333588-46593e80-f576-11e8-856e-df464faae940.png)

diosmosis commented:
@mattab UI needs approval

diosmosis commented on plugins/Transitions/angularjs/transitionswitcher/transitionswitcher.controller.js:
```diff
@@ -0,0 +1,104 @@
+                    self.isEnabled = false;
+                    self.actionName = '';
+                    self.actionNameOptions.push({key: '', value: translate('CoreHome_ThereIsNoDataForThisReport')});
+                }
```
Not super important, but I think this method could be split up for better readability, if desired.

tsteur replied:
will do 👍

tsteur commented on plugins/Transitions/lang/en.json:
```diff
@@ -2,11 +2,14 @@
         "NumDownloads": "%s downloads",
         "NumOutlinks": "%s outlinks",
+        "TopX": "Top %s labels",
+        "AvailableInOtherReports": "Did you know? Transitions are also available as a row action in the following reports:",
```
will add it: 👍

tsteur committed:
apply review feedback

tsteur commented:
> the report doesn't seem to be centered, might look better if it was

I didn't manage to center it but I actually think it's fine like this anyway. Feel free to give it a try. It's so hard to do much there I think cause the transitions JS is quite complex and built for popup etc.

> If I click in the transitions report, I see a popup like this

had actually no idea it was clickable... i think the JS is not able to render 2 transition reports on the same page or it's trying to alter something. 
It should more or less work now. If there's still a problem I will need to disable the onclick functionality in this report.

> the select dropdown is under the report

should be fixed

also applied the other feedback @diosmosis

diosmosis commented:
LGTM just need UX approval from @mattab & UI tests

tsteur committed:
Merge branch '3.x-dev' into 12865

tsteur committed:
add ui test, fix widget view

tsteur committed:
fix ui test

tsteur committed:
fix tests

tsteur commented:
Tests should be fixed now.

diosmosis committed:
Merge branch '3.x-dev' into 12865

diosmosis committed:
update screenshots

diosmosis commented:
Was going to merge, but the transitions widget looks a bit off:

![image](https://user-images.githubusercontent.com/125140/49691571-6484e880-fafa-11e8-929c-5804db03f921.png)

I think something in the code is also causing the UI test on travis to have issues (works fine for me locally): https://builds-artifacts.matomo.org/matomo-org/matomo/12865/31592/UIIntegrationTest_dashboard3.png

Think it might have something to do w/ the width of an element.

tsteur committed:
fix various issues

tsteur committed:
Merge branch '12865' of github.com:matomo-org/piwik into 12865

tsteur committed:
trying to fix tests

tsteur committed:
fix tests

diosmosis committed:
Update expected screenshot tests.

diosmosis committed:
Merge branch '3.x-dev' into 12865

tsteur commented:
@diosmosis I think I fixed the remaining issues this morning

diosmosis committed:
Try to fix random failure & update screenshots.

mattab commented:
Looks really good overall, it's nice to know more users will discover this feature.
Feedback:

* In the footer message eg. ` Simply hover a row in any of these reports and click on the transition icon to launch it. ` could maybe add the Transitions icon in the text itself.
* Add a "Help text" for the report, that 1) gives a quick explanation of what the report is, and 2) links to: https://matomo.org/docs/transitions/ 
* When viewing the transitions for a site, and changing site, then I get error:

```
Please specify a value for 'actionName'.#0 /core/API/Proxy.php(157): Piwik\API\Proxy->getRequestParametersArray(Array, Array)#1 /core/Context.php(28): Piwik\API\Proxy->Piwik\API\{closure}()#2 /core/API/Proxy.php(323): Piwik\Context::executeWithQueryParameters(Array, Object(Closure))#3 /core/API/Request.php(263): Piwik\API\Proxy->call('\\Piwik\\Plugins\\...', 'getTransitionsF...', Array)#4 /plugins/API/Controller.php(41): Piwik\API\Request->process()#5 [internal function]: Piwik\Plugins\API\Controller->index()#6 /core/FrontController.php(564): call_user_func_array(Array, Array)#7 /core/FrontController.php(152): Piwik\FrontController->doDispatch('API', false, Array)#8 /core/dispatch.php(34): Piwik\FrontController->dispatch()#9 /index.php(27): require_once('/home/matt/dev/...')#10 {main}
``` 

* in case it's possible, would be nice if we could re-draw the Transitions report when the browser window is resized (currently the Transitions report is not re-drawn and will overflow or look small)

tsteur committed:
add help text, add icon, do not fetch anything when no action name

tsteur commented:
* added the icon
* added help text & link
* couldn't reproduce an error when switching site.... added some code maybe it helps
* No we cannot resize it, it's a "miracle" it was possible to embed it at all

tsteur committed:
trying to fix tests

tsteur committed:
tweak code

tsteur committed:
run tests again

diosmosis committed:
Merge branch '3.x-dev' into 12865

diosmosis committed:
Update expected screenshots.
